crates/topos-tce-storage/src/tests/checkpoints.rs

```diff
@@ -0,0 +1,123 @@
+use std::{collections::HashMap, sync::Arc};
+
+use rstest::rstest;
+use topos_core::uci::SubnetId;
+use topos_test_sdk::{
+    certificates::create_certificate_chain,
+    constants::{SOURCE_SUBNET_ID_1, SOURCE_SUBNET_ID_2, TARGET_SUBNET_ID_1},
+};
+
+use super::support::store;
+use crate::{
+    store::{ReadStore, WriteStore},
+    validator::ValidatorStore,
+};
+
+#[rstest]
+#[tokio::test]
+async fn get_checkpoint_for_two_subnets(store: Arc<ValidatorStore>) {
+    let certificates_a = create_certificate_chain(SOURCE_SUBNET_ID_1, &[TARGET_SUBNET_ID_1], 32);
+    let certificates_b = create_certificate_chain(SOURCE_SUBNET_ID_2, &[TARGET_SUBNET_ID_1], 24);
+
+    for cert in certificates_a {
+        _ = store.insert_certificate_delivered(&cert).await;
+    }
+
+    for cert in certificates_b {
+        _ = store.insert_certificate_delivered(&cert).await;
+    }
+
+    let checkpoint = store
+        .get_checkpoint()
+        .unwrap()
+        .into_iter()
+        .map(|(subnet, value)| (subnet, *value.position))
+        .collect::<HashMap<SubnetId, u64>>();
+
+    assert_eq!(checkpoint.len(), 2);
+    assert_eq!(*checkpoint.get(&SOURCE_SUBNET_ID_1).unwrap(), 31);
+    assert_eq!(*checkpoint.get(&SOURCE_SUBNET_ID_2).unwrap(), 23);
+}
+
+#[rstest]
+#[tokio::test]
+async fn get_checkpoint_diff_with_no_input(store: Arc<ValidatorStore>) {
+    let certificates_a = create_certificate_chain(SOURCE_SUBNET_ID_1, &[TARGET_SUBNET_ID_1], 32);
+    let certificates_b = create_certificate_chain(SOURCE_SUBNET_ID_2, &[TARGET_SUBNET_ID_1], 24);
+
+    for cert in certificates_a {
+        _ = store.insert_certificate_delivered(&cert).await;
+    }
+
+    for cert in certificates_b {
+        _ = store.insert_certificate_delivered(&cert).await;
+    }
+
+    let checkpoint = store
+        .get_checkpoint_diff(&[])
+        .unwrap()
+        .into_iter()
+        .map(|(subnet, proofs)| {
+            (
+                subnet,
+                proofs
+                    .iter()
+                    .map(|proof| *proof.delivery_position.position)
+                    .collect::<Vec<_>>(),
+            )
+        })
+        .collect::<HashMap<SubnetId, _>>();
+
+    assert_eq!(checkpoint.len(), 2);
+    assert_eq!(
+        *checkpoint.get(&SOURCE_SUBNET_ID_1).unwrap(),
+        (0..=31).collect::<Vec<_>>()
+    );
+    assert_eq!(
+        *checkpoint.get(&SOURCE_SUBNET_ID_2).unwrap(),
+        (0..=23).collect::<Vec<_>>()
+    );
+}
+
+#[rstest]
+#[tokio::test]
+async fn get_checkpoint_diff_with_input(store: Arc<ValidatorStore>) {
```

Comment:
I do not understand what this test actually does, can you explain please? The part I don't get is what the "input" part is. The only difference between this test and the "no_input" version seems to be that this one does `.get(20)` and clones the proof of delivery, but I don't see how that maps to "with input".

Reply:
In this test, the `get_checkpoint_diff` method has an input of `&[checkpoint]` (line 100) while the `get_checkpoint_diff_with_no_input` `get_checkpoint_diff` method doesn't have one (line 57).